runtime/compiler/control/CompilationThread.cpp

```diff
@@ -7036,7 +7077,48 @@ TR::CompilationInfoPerThreadBase::preCompilationTasks(J9VMThread * vmThread,
             }
 
          if (!doLocalCompilation)
-            entry->setRemoteCompReq();
+            {
+            JITaaS::J9ClientStream *client = getClientStream();
+            if (!client)
+               {
+               if (JITaaS::J9ClientStream::_serverReachable || JITaaS::J9ClientStream::_numLocalCompilations >= 50)
+                  {
+                  try
+                     {
+                     client = new (PERSISTENT_NEW) JITaaS::J9ClientStream(_compInfo.getPersistentInfo());
+                     setClientStream(client);
+                     entry->setRemoteCompReq();
+                     JITaaS::J9ClientStream::_serverReachable = true;
+                     }
+                  catch (const JITaaS::StreamFailure &e)
+                     {
+                     if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
+                        TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, e.what());
+                     }
+                  catch (const JITaaS::StreamUnreachable &e)
+                     {
+                     // failed to connect to JITaaS server, perform local compilation instead of remote compilation
+                     JITaaS::J9ClientStream::_numLocalCompilations = 0;
+                     JITaaS::J9ClientStream::_serverReachable = false;
+                     }
+                  catch (const std::bad_alloc &e)
+                     {
+                     if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
+                        TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, e.what());
+
+                     throw;
+                     }
+                  }
+               else
+                  {
+                  JITaaS::J9ClientStream::_numLocalCompilations++;
+                  }
+               }
+            else
+               {
+               entry->setRemoteCompReq();
+               }
+            }
```

Comment:
This chunk of code looks identical to the chunk above, is one for AOT and the other for JIT? Can you try to encapsulate this into a function rather than repeating it? It will save us headaches maintaining it in the future.